src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -907,45 +908,36 @@ public static int Sum<T>(this ImmutableArray<T> items, Func<T, int> selector)
 
         internal static void CreateNameToMembersMap
             <TNamespaceOrTypeSymbol, TNamedTypeSymbol, TNamespaceSymbol>
-            (Dictionary<string, object> dictionary, Dictionary<String, ImmutableArray<TNamespaceOrTypeSymbol>> result)
+            (Dictionary<string, object> dictionary, Dictionary<string, ImmutableArray<TNamespaceOrTypeSymbol>> result)
             where TNamespaceOrTypeSymbol : class
             where TNamedTypeSymbol : class, TNamespaceOrTypeSymbol
             where TNamespaceSymbol : class, TNamespaceOrTypeSymbol
         {
             foreach (var kvp in dictionary)
-            {
-                object value = kvp.Value;
-                ImmutableArray<TNamespaceOrTypeSymbol> members;
+                result.Add(kvp.Key, createMembers(kvp.Value));
+
+            return;
 
-                var builder = value as ArrayBuilder<TNamespaceOrTypeSymbol>;
-                if (builder != null)
+            static ImmutableArray<TNamespaceOrTypeSymbol> createMembers(object value)
+            {
+                if (value is ArrayBuilder<TNamespaceOrTypeSymbol> builder)
```

Comment:
no need for 'hasNamespaces' flag and complex data flow.  Once we see a namespace, we know we have to return the builder directly as an ImmutableArray of namespaces or types.